gandalf.js

```diff
@@ -1,41 +1,48 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?